### ISSUE TITLE
The parameters are not updated when the pruned model is finetuned

### ISSUE THREAD

**Issue by typhoonlee** [microsoft/nni#4521]
**Describe the issue**:
The parameters of the pruned model are not updated when finetune is performed, but the loss value can be calculated, as if the loss value is not back-propagated to update the parameters

**Environment**:
- NNI version: 2.6
- Training service (local|remote|pai|aml|etc): local
- Client OS: linux
- Server OS (for remote mode only): 
- Python version: 3.8
- PyTorch/TensorFlow version: torch 1.8
- Is conda/virtualenv/venv used?: conda
- Is running in Docker?: no

this is my code, I added different stages of pruning in train.py
`    for epoch in range(start_epoch, total_epochs + 1):
        train_sampler.set_epoch(epoch)
        prefetcher.reset()
        train_data = prefetcher.next()

        while train_data is not None:
            data_time = time.time() - data_time  # 加载数据的时间

            current_iter += 1
            if current_iter > total_iters:
                break
            # pruning model
            if opt['prune']:
                if current_iter == 100:
                    config_list = [{'sparsity': 0.25, 'op_types': ['Conv2d'], 'op_names': ['conv_first']},
                                   {'sparsity': 0.25, 'op_types': ['Conv2d'], 'op_names': ['body.0.conv1']},
                                   {'sparsity': 0.25, 'op_types': ['Conv2d'], 'op_names': ['body.0.conv2']},
                                   {'sparsity': 0.25, 'op_types': ['Conv2d'], 'op_names': ['body.0.conv3']}]
                    pruner = L1FilterPruner(model.net_g, config_list)
                    pruner.compress()
                    pruner.export_model(model_path='pruned_naive_mnist_l1filter.pth',
                                        mask_path='mask_naive_mnist_l1filter.pth')
                    pruner._unwrap_model()
                    m_speedup = ModelSpeedup(model.net_g, dummy_input=torch.rand(1, 1, 28, 28).cuda(),
                                             masks_file='mask_naive_mnist_l1filter.pth')
                    m_speedup.speedup_model()
                elif current_iter == 5000:
                    config_list = [{'sparsity': 0.34, 'op_types': ['Conv2d'], 'op_names': ['conv_first']},
                                   {'sparsity': 0.34, 'op_types': ['Conv2d'], 'op_names': ['body.0.conv1']},
                                   {'sparsity': 0.34, 'op_types': ['Conv2d'], 'op_names': ['body.0.conv2']},
                                   {'sparsity': 0.34, 'op_types': ['Conv2d'], 'op_names': ['body.0.conv3']}]
                    pruner = L1FilterPruner(model.net_g, config_list)
                    pruner.compress()
                    pruner.export_model(model_path='pruned_naive_mnist_l1filter.pth',
                                        mask_path='mask_naive_mnist_l1filter.pth')
                    pruner._unwrap_model()
                    m_speedup = ModelSpeedup(model.net_g, dummy_input=torch.rand(1, 1, 28, 28,requires_grad=True).cuda(),
                                             masks_file='mask_naive_mnist_l1filter.pth')
                    m_speedup.speedup_model()
            # update learning rate
            model.update_learning_rate(current_iter, warmup_iter=opt['train'].get('warmup_iter', -1))
            # training
            model.feed_data(train_data)
            model.optimize_parameters(current_iter)
            iter_time = time.time() - iter_time      # 每个iter的时间
            # log

            if current_iter % opt['logger']['print_freq'] == 0:
                log_vars = {'epoch': epoch, 'iter': current_iter}
                log_vars.update({'lrs': model.get_current_learning_rate()})
                log_vars.update({'time': iter_time, 'data_time': data_time})
                log_vars.update(model.get_current_log())
                msg_logger(log_vars)

            # save models and training states
            if current_iter % opt['logger']['save_checkpoint_freq'] == 0:
                logger.info('Saving models and training states.')
                model.save(epoch, current_iter)

            # validation
            # if opt.get('val') is not None and (current_iter % opt['val']['val_freq'] == 0):
            #     model.validation(val_loader, current_iter, tb_logger, opt['val']['save_img'])

            data_time = time.time()
            iter_time = time.time()
            train_data = prefetcher.next()`

and this is error：
/home/amlogic/anaconda3/envs/SR/lib/python3.8/site-packages/nni/compression/pytorch/speedup/compressor.py:282: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the gradient for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations.
  if last_output.grad is not None and tin.grad is not None:

**How to reproduce it?**:

**Comment by J-shang** [comment]
Hi @typhoonlee , I think this may be because of the wrong usage of `requires_grad`, remove `requires_grad=True`, and have a try.

FYI, https://discuss.pytorch.org/t/grad-attribute-of-a-non-leaf-tensor-being-accessed/82313

**Comment by typhoonlee** [comment]
> 
Thanks for your reply, but this error appears when iter==100, specifically in `m_speedup = ModelSpeedup(model.net_g, dummy_input=torch.rand(1, 1, 28, 28).cuda(),masks_file='mask_naive_mnist_l1filter.pth')` appears here, and does not use requires_grad=True
And after iter==100 pruning, the resulting new model will no longer update the weight in subsequent training.

**Comment by J-shang** [comment]
> > 
> 
> Thanks for your reply, but this error appears when iter==100, specifically in `m_speedup = ModelSpeedup(model.net_g, dummy_input=torch.rand(1, 1, 28, 28).cuda(),masks_file='mask_naive_mnist_l1filter.pth')` appears here, and does not use requires_grad=True And after iter==100 pruning, the resulting new model will no longer update the weight in subsequent training.

1) Could you give me an example that I can reproduce this error?
2) `m_speedup.speedup_model()` will replace the layers in the original model with smaller ones, then copy the `weight` values (un-masked part) to the smaller layers. So the parameters registered in the optimizer are outdated. You need to update the parameters in the optimizer to the new ones, an easy way is reinitializing a new optimizer.

**Comment by typhoonlee** [comment]
> > > 
> > 
> > 
> > Thanks for your reply, but this error appears when iter==100, specifically in `m_speedup = ModelSpeedup(model.net_g, dummy_input=torch.rand(1, 1, 28, 28).cuda(),masks_file='mask_naive_mnist_l1filter.pth')` appears here, and does not use requires_grad=True And after iter==100 pruning, the resulting new model will no longer update the weight in subsequent training.
> 
> 1. Could you give me an example that I can reproduce this error?
> 2. `m_speedup.speedup_model()` will replace the layers in the original model with smaller ones, then copy the `weight` values (un-masked part) to the smaller layers. So the parameters registered in the optimizer are outdated. You need to update the parameters in the optimizer to the new ones, an easy way is reinitializing a new optimizer.

thank you very much! I solved this problem by redefining an optimizer
